Update/enforce jetpack modules setup

paulopmt1 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Slack thread: p1744052983694779-slack-CDLH4C1UZ

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* We're enforcing the activation of three Jetpack modules that are getting disabled after a Commerce upgrade

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p1744052983694779-slack-CDLH4C1UZ

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
no

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Create a simple site and add it to Atomic Dev Site
* Transfer it to Atomic
* Upgrade the site to Commerce plan
* Once the its on Commerce, open the editor and make sure the Jetpack blocks are disabled
* Also run `wp jetpack module list` to make sure the three modules are disabled:
	![image](https://github.com/user-attachments/assets/e1b8db7d-7c60-4f54-aa99-c240e22ed7e8)

* Install [Jetpack beta tester plugin](https://jetpack.com/download-jetpack-beta/) on the site you just created
* Open the plugin page and select the `WordPress.com Site Helper`
* Install the current PR `update/enforce-jetpack-modules-setup` on the site
* Run the `wp wpcomsh post-transfer` command to simulate an Atomic transfer and see that after another `jetpack module list` command, the three modules will be enabled:

	![image](https://github.com/user-attachments/assets/54a719cd-5ada-4ca9-9c49-81293ed35d3e)

* Jetpack modules will be enabled again:

	![image](https://github.com/user-attachments/assets/10e5bab2-074a-454c-9059-e4980185e98e)

paulopmt1 committed:
Enforce jetpack modules after Atomic transfer

paulopmt1 committed:
changelog

paulopmt1 committed:
Added post-reset trigger for the same enforcing fucntion

jeherve commented on projects/plugins/wpcomsh/woa.php:
```diff
@@ -393,3 +393,65 @@ function wpcomsh_woa_post_process_store_woocommerce_connection_details( $args, $
+		return;
+	}
+
+	// First, make sure the jetpack_blocks_disabled option is deleted
+	delete_option( 'jetpack_blocks_disabled' );
```
Do we need to delete that option when it's done later when the module gets activated?
https://github.com/Automattic/jetpack/blob/9c556174eee294544a57d736d095ef05c8539d8e/projects/plugins/jetpack/modules/blocks.php#L15-L30

kraftbj replied:
I suppose it wouldn't hurt if the block is already active and had been set without the activation flow.

jeherve commented on projects/plugins/wpcomsh/woa.php:
```diff
@@ -393,3 +393,65 @@ function wpcomsh_woa_post_process_store_woocommerce_connection_details( $args, $
+	WP_CLI::log( 'Currently active Jetpack modules:' );
+	WP_CLI::log( $active_modules_result->stdout );
+
+	WP_CLI::success( 'Jetpack modules activation completed' );
```
Should we only run this if we don't get a warning of a failed activation above?

paulopmt1 replied:
Good catch, [fixed it here](https://github.com/Automattic/jetpack/pull/42971/commits/10f2b225880c4b641f140f82702ca7c0957b5e9a).

jeherve commented on projects/plugins/wpcomsh/woa.php:
```diff
@@ -393,3 +393,65 @@ function wpcomsh_woa_post_process_store_woocommerce_connection_details( $args, $
+		}
+	}
+
+	// Get a list of all active modules to verify
+	$active_modules_result = WP_CLI::runcommand(
+		'jetpack module list --status=active',
+		array(
+			'return'     => 'all',
+			'launch'     => false,
+			'exit_error' => false,
+		)
+	);
+
+	WP_CLI::log( 'Currently active Jetpack modules:' );
+	WP_CLI::log( $active_modules_result->stdout );
```
I'm not familiar with the transfer process, but is that here because it's logged somewhere so we have access to that data for debugging later?

paulopmt1 replied:
Good question. As I'm also not familiar with it, I'm reusing it since we're using it for all activations above in that file. 

I did check the site-logs Hosting interface, and the logs weren't there. I'd assume we can access them at the log stash. I'll check that once we land the code, because since I haven't found them yet, I suspect we won't have sandbox logs there.

kraftbj commented on projects/plugins/wpcomsh/woa.php:
```diff
@@ -393,3 +393,65 @@ function wpcomsh_woa_post_process_store_woocommerce_connection_details( $args, $
+		'blocks',
+		'account-protection',
+		'blaze',
+	);
```
```suggestion
	$modules_to_activate = array(
		'account-protection',
		'blaze',
		'blocks',
	);
```

Super nit-picky, but putting it into alpha sort while we're looking at this.

paulopmt1 replied:
Cool, [fixed it here](https://github.com/Automattic/jetpack/pull/42971/commits/10f2b225880c4b641f140f82702ca7c0957b5e9a).

kraftbj reviewed (changes requested):
Agree with @jeherve's review. The 452 comment is the only one that I'd consider a blocker.

Perhaps something like `$success = null` value that `if ( $success !== false ) { $success = true )` when it passes or `$success = false` when it doesn't, then output a final `if success, `etc)

paulopmt1 commented:
Thanks for all the feedback, folks! 
I was focused on releasing a project today and will be AFK tomorrow. Following Monday, I'll get back to it.

paulopmt1 committed:
Code improvements:
- Improved module order
- Only showing activation completed when all modules were enabled

paulopmt1 commented:
> The 452 comment is the only one that I'd consider a blocker.

paulopmt1 commented:
> The 452 comment is the only one that I'd consider a blocker.

Thanks for the review @kraftbj, I fixed this blocker [issue here](https://github.com/Automattic/jetpack/pull/42971/commits/10f2b225880c4b641f140f82702ca7c0957b5e9a).

kraftbj reviewed (approved):
Looks good to me. I reviewed the module activation code and confirmed that we return `true` when we either activate a module or request a module that is already active.
